website/notifications/events/utils.py

```diff
@@ -0,0 +1,124 @@
+    :param node: Node instance to
+    :return: move, warn, and remove dicts
+    """
+    # initialise subscription dictionaries
+    move = {key: [] for key in constants.NOTIFICATION_TYPES}
+    warn = {key: [] for key in constants.NOTIFICATION_TYPES}
+    remove = {key: [] for key in constants.NOTIFICATION_TYPES}
+    # get the node subscription
+    if len(files) == 0:
+        move, warn, remove = categorize_users(
+            user, 'file_updated', source_node, 'file_updated', node
+        )
+    # iterate through file subscriptions
+    for file_path in files:
+        path = file_path.strip('/')
+        t_move, t_warn, t_remove = categorize_users(
+            user, path + '_file_updated', source_node,
+            path + '_file_updated', node
+        )
+        # Add file subs to overall list of subscriptions
+        for notification in constants.NOTIFICATION_TYPES:
+            move[notification] = list(set(move[notification]).union(set(t_move[notification])))
+            warn[notification] = list(set(warn[notification]).union(set(t_warn[notification])))
+            remove[notification] = list(set(remove[notification]).union(set(t_remove[notification])))
+    return move, warn, remove
+
+
+def categorize_users(user, source_event, source_node, event, node):
+    """Categorize users from a file subscription into three categories.
+
+    Puts users in one of three bins: Those that are moved, those that
+     need warned, those that are removed.
+    Calls move_subscription in order to move the sub and get users w/o permissions
+    :param user: User instance who started the event
+    :param source_event: <guid>_event_name
+    :param source_node: node from where the event happened
+    :param event: new guid event name
+    :param node: node where event ends up
+    :return: Moved, to be warned, and removed users.
+    """
+    remove = utils.users_to_remove(source_event, source_node, node)
+    source_node_subs = compile_subscriptions(source_node, utils.find_subscription_type(source_event))
+    new_subs = compile_subscriptions(node, utils.find_subscription_type(source_event), event)
+    warn = {key: [] for key in constants.NOTIFICATION_TYPES}
+    move = {key: [] for key in constants.NOTIFICATION_TYPES}
+
+    # Set operations to separate users
+    for notifications in constants.NOTIFICATION_TYPES:
+        if notifications == 'none':
+            continue
+        move[notifications] = list(set(source_node_subs[notifications]).union(set(new_subs[notifications])))
+        warn[notifications] = list(set(source_node_subs[notifications]).difference(set(new_subs[notifications])))
+        subbed, removed = utils.separate_users(node, warn[notifications])
+        warn[notifications] = subbed
+        remove[notifications].extend(removed)
+        remove[notifications] = list(set(remove[notifications]))
+
+    # Remove duplicates in different types
+    for notifications in constants.NOTIFICATION_TYPES:
```

Comment:
Is it necessary to loop over the same dictionary 5 times? 

Before I review this in-depth, please review this and consider opportunities for refactoring.